### ISSUE TITLE
Memory leak in hipCreateTextureObject and hipDestroyTextureObject (hipErrorOutOfMemory)

### ISSUE THREAD

**Issue by ex-rzr** [ROCm-Developer-Tools/HIP#2296]
The original program recreates a texture every few iterations (we needed to add a workaround with swapping two textures instead of recreating one).

Here is a simplified test case:

```cpp
#include <cstdio>

#include <hip/hip_runtime.h>
#include <hip/hip_runtime_api.h>

// hipcc tex_mem_leak.cpp -o tex_mem_leak && ./tex_mem_leak

int main() {
    const unsigned int size = 1024;

    float * buffer;
    hipMalloc(&buffer, size * sizeof(float));

    hipError_t status;

    for (int i = 0; i < 100000; i++) {
        printf("iteration: %4d\n", i);
        hipTextureObject_t tex;
        hipResourceDesc resDesc;
        hipTextureDesc texDesc;
        memset(&resDesc, 0, sizeof(resDesc));
        resDesc.resType = hipResourceTypeLinear;
        resDesc.res.linear.devPtr = buffer;
        resDesc.res.linear.sizeInBytes = size * sizeof(float);
        resDesc.res.linear.desc.f = hipChannelFormatKindFloat;
        resDesc.res.linear.desc.x = 32;
        resDesc.res.linear.desc.y = 0;
        resDesc.res.linear.desc.z = 0;
        resDesc.res.linear.desc.w = 0;
        memset(&texDesc, 0, sizeof(texDesc));
        texDesc.readMode = hipReadModeElementType;
        status = hipCreateTextureObject(&tex, &resDesc, &texDesc, NULL);
        if (status) {
            fprintf(stderr, "hipCreateTextureObject failed: %d\n", (int)status);
            exit(-1);
        }

        hipDeviceSynchronize();

        status = hipDestroyTextureObject(tex);
        if (status) {
            fprintf(stderr, "hipDestroyTextureObject failed: %d\n", (int)status);
            exit(-1);
        }
    }

    hipDeviceSynchronize();
    hipFree(buffer);
}
```
[tex_mem_leak.zip](https://github.com/ROCm-Developer-Tools/HIP/files/6691131/tex_mem_leak.zip)

Build and run it with:
```
hipcc tex_mem_leak.cpp -o tex_mem_leak && ./tex_mem_leak
```

It fails at the same iteration:
```
iteration: 65353
Depth set on single layer image geometry.
hipCreateTextureObject failed: 2
```

With `AMD_LOG_LEVEL=4`:
```
iteration: 65352
:3:hip_texture.cpp          :315 : 2143524677122 us: 1691988: [7feea7b5a840] hipCreateTextureObject ( 0x7ffec348d5e8, {2,{0x7feda5400000,{32,0,0,0,hipChannelFormatKindFloat},4096}, {{hipAddressModeWrap,hipAddressModeWrap,hipAddressModeWrap},hipFilterModePoint,hipReadModeElementType,0,{0,0,0,0},0,hipFilterModePoint,0,0,0}, char array:<null> )
:3:rocdevice.cpp            :2065: 2143524677197 us: device=0x2035740, freeMem_ = 0xef0b6000
Depth set on single layer image geometry.
:4:rocdevice.cpp            :1890: 2143524677301 us: Allocate hsa host memory 0x7fe86dfc0000, size 0x110
:3:hip_texture.cpp          :317 : 2143524677486 us: 1691988: [7feea7b5a840] hipCreateTextureObject: Returned hipSuccess : 
:3:hip_device_runtime.cpp   :458 : 2143524677492 us: 1691988: [7feea7b5a840] hipDeviceSynchronize (  )
:4:commandqueue.cpp         :119 : 2143524677501 us: All commands finished
:3:hip_device_runtime.cpp   :470 : 2143524677506 us: 1691988: [7feea7b5a840] hipDeviceSynchronize: Returned hipSuccess : 
:3:hip_texture.cpp          :339 : 2143524677516 us: 1691988: [7feea7b5a840] hipDestroyTextureObject ( 0x7fe86dfc0000 )
:3:hip_memory.cpp           :322 : 2143524677546 us: 1691988: [7feea7b5a840] hipFree ( 0x7fe86dfc0000 )
:4:commandqueue.cpp         :119 : 2143524677552 us: All commands finished
:4:rocdevice.cpp            :2042: 2143524677586 us: Free hsa memory 0x7fe86dfc0000
:3:hip_memory.cpp           :324 : 2143524677594 us: 1691988: [7feea7b5a840] hipFree: Returned hipSuccess : 
:3:hip_texture.cpp          :341 : 2143524677600 us: 1691988: [7feea7b5a840] hipDestroyTextureObject: Returned hipSuccess : 
iteration: 65353
:3:hip_texture.cpp          :315 : 2143524677625 us: 1691988: [7feea7b5a840] hipCreateTextureObject ( 0x7ffec348d5e8, {2,{0x7feda5400000,{32,0,0,0,hipChannelFormatKindFloat},4096}, {{hipAddressModeWrap,hipAddressModeWrap,hipAddressModeWrap},hipFilterModePoint,hipReadModeElementType,0,{0,0,0,0},0,hipFilterModePoint,0,0,0}, char array:<null> )
:3:rocdevice.cpp            :2065: 2143524677701 us: device=0x2035740, freeMem_ = 0xef0b5000
Depth set on single layer image geometry.
:4:rocdevice.cpp            :1890: 2143524677795 us: Allocate hsa host memory (nil), size 0x110
:1:rocdevice.cpp            :1892: 2143524677801 us: Fail allocation host memory
:1:rocdevice.cpp            :1786: 2143524677806 us: Failed creating memory
:1:memory.cpp               :337 : 2143524677811 us: Video memory allocation failed!
:1:memory.cpp               :1489: 2143524677817 us: Unable to allocate aligned memory
:3:hip_memory.cpp           :408 : 2143524677828 us: 1691988: [7feea7b5a840] hipMemGetInfo ( 0x7ffec348d4a8, 0x7ffec348d4b0 )
:3:hip_memory.cpp           :423 : 2143524677836 us: 1691988: [7feea7b5a840] hipMemGetInfo: Returned hipSuccess : 
:1:hip_memory.cpp           :136 : 2143524677842 us: Allocation failed : Device memory : required :272 | free :34075267072 | total :34342961152 

:3:hip_texture.cpp          :317 : 2143524677850 us: 1691988: [7feea7b5a840] hipCreateTextureObject: Returned hipErrorOutOfMemory : 
hipCreateTextureObject failed: 2

```

**Comment by vsytch** [comment]
Thank you for reporting this issue. I was able to reproduce this issue on Linux, but not on Windows, which is strange, because the texture allocation/release logic is OS independent.

![test_mem_leak_win](https://user-images.githubusercontent.com/22281896/128066829-eae1dd6c-74c5-4368-bd37-dddb39b7c8e2.png)
![test_mem_leak_lnx](https://user-images.githubusercontent.com/22281896/128066848-867dac27-7f9d-467c-a142-b1429556e270.png)

**Comment by ex-rzr** [comment]
Perhaps the runtime allocates a larger buffer for textures on Windows? (hence more iterations are required)

**Comment by vsytch** [comment]
We forgot to free the HW sampler... Not sure how this test was passing on Windows.
![image](https://user-images.githubusercontent.com/22281896/128616001-e42a110c-42dc-45dc-86a0-318ad9ecb45d.png)

**Comment by gandryey** [comment]
On windows sampler SRDs are coming from a suballoc and it's tiny. ROCr doesn't support suballocs and each allocation will consume much bigger chunk of memory due to the alignment, like 100x bigger. That's why you didn't notice the issue under windows.

**Comment by vsytch** [comment]
Fix was submitted internally. Expect it to be available with next ROCm release. I guess that would be 4.4?

In the meanwhile, if you need this fix, please apply this patch [66ab78a.diff.zip](https://github.com/ROCm-Developer-Tools/HIP/files/6958386/66ab78a.diff.zip) to https://github.com/ROCm-Developer-Tools/hipamd and rebuild the hip runtime.

**Comment by mangupta** [comment]
Closing based on last comment.
